### PR TITLE
Bayeux protocol (cometD) client implementation with Salesforce example

### PR DESCRIPTION
**Pull request by ragrus-nbl**
Final testing is in progress.

### CODE REVIEW

File: BayeuxClient.agent.lib.nut [hunk 1]
```diff
@@ -858,13 +853,15 @@ class Bayeux.Client {
...
-        foreach (k, v in response.headers) {
-            if (k.tolower() == "set-cookie") {
+        foreach (header in response.rawheaders) {
+            if (header.k.tolower() == "set-cookie") {
                 try {
-                    local cookieName = split(v, "=")[0];
-                    _cookies[cookieName] <- v;
+                    // Cookie's value should be between the first "=" and the first ";"
+                    local startIdx = header.v.find("=");
+                    local endIdx = header.v.find(";");
```
Comment (ppetrosh) [inline BayeuxClient.agent.lib.nut]: what if the cookie looks like this:

```
Set-Cookie: theme=light
```

(seems to be a valid cookie for me)

File: BayeuxClient.agent.lib.nut [hunk 2]
```diff
@@ -858,13 +853,15 @@ class Bayeux.Client {
...
-        // Waiting for EI to fix the issue with Set-Cookie headers
-        foreach (k, v in response.headers) {
-            if (k.tolower() == "set-cookie") {
+        foreach (header in response.rawheaders) {
+            if (header.k.tolower() == "set-cookie") {
                 try {
-                    local cookieName = split(v, "=")[0];
-                    _cookies[cookieName] <- v;
+                    // Cookie's value should be between the first "=" and the first ";"
+                    local startIdx = header.v.find("=");
```
Comment (ppetrosh) [inline BayeuxClient.agent.lib.nut]: does the standard allow for spaces around `=`?

File: BayeuxClient.agent.lib.nut (hunk 1 above)
Comment (ragrus-nbl) [reply BayeuxClient.agent.lib.nut]: Yes, fixed

File: BayeuxClient.agent.lib.nut (hunk 2 above)
Comment (ragrus-nbl) [reply BayeuxClient.agent.lib.nut]: Spaces are forbidden but now we handle this, just in case

File: BayeuxClient.agent.lib.nut [hunk 3]
```diff
@@ -0,0 +1,961 @@
...
+                    if (startIdx == null) {
+                        continue;
+                    }
+
+                    cookieName = strip(header.v.slice(0, startIdx));
+                    if (cookieName == "") {
+                        continue;
+                    }
+
+                    if (endIdx == null) {
```
Comment (ppetrosh) [inline BayeuxClient.agent.lib.nut:902]:
```
                    if (endIdx == null) {
                        endIdx = header.v.len();
                    }

                    value = strip(header.v.slice(startIdx + 1, endIdx));
```

File: examples/Salesforce.agent.nut [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO
+// EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES
+// OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+// ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+// OTHER DEALINGS IN THE SOFTWARE.
+
+// Utility Libraries
+#require "Rocky.class.nut:1.2.3"
```
Comment (ppetrosh) [inline examples/Salesforce.agent.nut]: Why are you using that old version?

File: BayeuxClient.agent.lib.nut [hunk 5]
```diff
@@ -0,0 +1,961 @@
...
+
+        if (tableMsg != null) {
+            callback(null);
+            _handleMessages(tableMsg);
+        }
+
+        _processQueue();
+    }
+
+    function _updateCookies(headers) {
```
Comment (ppetrosh) [inline BayeuxClient.agent.lib.nut:877]: Just curious, should we ever clean cookies up?

File: BayeuxClient.agent.lib.nut (hunk 3 above)
Comment (ragrus-nbl) [reply BayeuxClient.agent.lib.nut:902]: Done

File: examples/Salesforce.agent.nut (hunk 4 above)
Comment (ragrus-nbl) [reply examples/Salesforce.agent.nut]: Updated

File: BayeuxClient.agent.lib.nut (hunk 5 above)
Comment (ragrus-nbl) [reply BayeuxClient.agent.lib.nut:877]: We clean them after every disconnection. But there might be some cases where we should also clean them. Added a TODO on it.

**Pull request merged**